astropy/io/fits/tests/test_core.py

```diff
@@ -996,6 +997,49 @@ def test_filename_with_colon(self):
         with fits.open(self.temp(filename)) as hdul:
             assert np.all(hdul[0].data == hdu.data)
 
+
+    def test_writeto_full_disk(self, monkeypatch):
+        """
+        Test that it gives a readable error when trying to write an hdulist
+        to a full disk.
+        """
+        def _writeto(self, array):
+            raise IOError("Fake error raised for writing to files")
+
+        with pytest.raises(IOError) as e:
+            monkeypatch.setattr(fits.hdu.base._BaseHDU, "_writeto", _writeto)
+
+            filename = self.temp('test.fits')
+            hdul = [fits.PrimaryHDU(), fits.ImageHDU()]
+            hdul = fits.HDUList(hdul)
+            hdul.writeto(filename)
+            hdul.close()
+
+            assert "Fake error raised for writing to files" in e.value
```

Comment:
This test is slightly flaky in that because it runs `get_free_space_in_dir` that *could* actually fail, and instead of getting the original IOError you raised you could end up getting the actual "Not enough space" error instead, which would fail the test (and yet ironically is the feature we're more interested in testing.  

I think you should also monkeypatch `get_free_space_in_dir` itself, which we don't care about testing for the purpose of this test.  Just change it to return `0` so that the "Not enough space on disk. " error always gets raised and we can test for that, since that's the actual use case this change was designed to handle (as opposed to some other arbitrary `IOError` being handled or not handled).